bot/conversation.py

```diff
@@ -257,23 +224,107 @@ async def heating(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
             inline_keyboard_button("Любое", "all", data),
         ],
         [
-            InlineKeyboardButton("Подтвердить", callback_data="confirm"),
+            InlineKeyboardButton("Назад", callback_data="_back"),
         ],
     ]
     await update.callback_query.answer()
     text = "Выбери тип отопления"
-    if callback_data == "heating":
-        await context.bot.send_message(
-            user_id,
-            text,
-            reply_markup=InlineKeyboardMarkup(reply_keyboard),
-        )
+    await update.callback_query.edit_message_text(
+        text,
+        reply_markup=InlineKeyboardMarkup(reply_keyboard),
+    )
+    return CHECK_HEATING
+
+
+async def get_towns(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
+    towns = db.towns.find()
+    context.user_data["areas"] = context.user_data.get("areas", {})
+
+    reply_keyboard = [[]]
+    for town in towns:
+        reply_keyboard[-1].append(InlineKeyboardButton(town["name"], callback_data=town["_id"]))
+        if not "all_" + town["_id"] in context.user_data["areas"]:
+            context.user_data["areas"]["all_" + town["_id"]] = False
+    reply_keyboard.append([InlineKeyboardButton("Назад", callback_data="_back")])
+    logging.info(context.user_data["areas"])
+    await update.callback_query.answer()
+    text = "Выбери город"
+    await update.callback_query.edit_message_text(
+        text,
+        reply_markup=InlineKeyboardMarkup(reply_keyboard),
+    )
+    return AREAS
+
+
+async def get_areas(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
+    town_id, area, *_ = update.callback_query.data.split('&') + ['', '']
+    areas = {area["name"]: False for area in db.areas.find({"town_id": town_id}).sort("name", 1)}
+    if not area:
+        context.user_data["areas"]["all_" + town_id] = True
+        context.user_data["areas"].update(areas)
+    elif area == "all":
+        context.user_data["areas"]["all_" + town_id] = not context.user_data["areas"]["all_" + town_id]
+        context.user_data["areas"].update(areas)
```

Comment:
не напрягает везде писать длинное `context.user_data["..."]` ?

 ```suggestion
    town_id, area, *_ = update.callback_query.data.split('&') + ['', '']
    areas = {area["name"]: False for area in db.areas.find({"town_id": town_id}).sort("name", 1)}
    user_areas = context.user_data["areas"]
    if not area:
        user_areas["all_" + town_id] = True
        user_areas.update(areas)
    elif area == "all":
        user_areas["all_" + town_id] = not context.user_data["areas"]["all_" + town_id]
        user_areas.update(areas)
```

Reply:
вообще как-то странно код смотрится, мне кажется можно сократить.
и префиксы "all_" странные, может проще складывать сразу в `context.user_data["towns"][town_id]` ?

Reply:
Напрягает, думал присвоить короткую ссылку

Reply:
> вообще как-то странно код смотрится, мне кажется можно сократить. и префиксы "all_" странные, может проще складывать сразу в `context.user_data["towns"][town_id]` ?

Я так сначала думал, но выбрал этот вариант.